stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -1750,34 +1754,53 @@ impl NakamotoChainState {
         sortdb: &SortitionDB,
         sort_handle: &SortitionHandleConn,
         chainstate: &mut StacksChainState,
-        for_block_height: u64,
+        for_burn_block_height: u64,
         at_block_id: &StacksBlockId,
     ) -> Result<Point, ChainstateError> {
         // Get the current reward cycle
         let Some(reward_cycle) = sort_handle
             .context
             .pox_constants
-            .block_height_to_reward_cycle(sort_handle.context.first_block_height, for_block_height)
+            .block_height_to_reward_cycle(
+                sort_handle.context.first_block_height,
+                for_burn_block_height,
+            )
         else {
             // This should be unreachable, but we'll return an error just in case.
             let msg = format!(
-                "BUG: Failed to determine reward cycle of block height: {}.",
-                for_block_height
+                "BUG: Failed to determine reward cycle of burn block height: {}.",
+                for_burn_block_height
             );
             warn!("{msg}");
             return Err(ChainstateError::InvalidStacksBlock(msg));
         };
 
-        chainstate
-            .get_aggregate_public_key_pox_4(sortdb, at_block_id, reward_cycle)?
-            .ok_or_else(|| {
-                warn!(
-                    "Failed to get aggregate public key";
-                    "block_id" => %at_block_id,
-                    "reward_cycle" => reward_cycle,
-                );
-                ChainstateError::InvalidStacksBlock("Failed to get aggregate public key".into())
-            })
+        // need to search back because the set-aggregate-public-key call only happens in nakamoto
+        // TODO: this will be removed once there's aggregate public key voting
+        for rc in (0..=reward_cycle).rev() {
+            debug!("get-aggregate-public-key {} {}", at_block_id, rc);
+            match chainstate.get_aggregate_public_key_pox_4(sortdb, at_block_id, rc)? {
+                Some(agg_key) => {
+                    return Ok(agg_key);
+                }
+                None => {
+                    debug!(
+                        "No aggregate public key set; trying in a lower cycle";
+                        "block_id" => %at_block_id,
+                        "reward_cycle" => rc,
+                    );
+                    continue;
+                }
+            }
+        }
```

Comment:
We call `NakamotoChainState::get_aggregate_public_key()` prior to processing the first Nakamoto block (which contains the `set-aggregate-public-key` call), since the relayer needs the aggregate public key to validate the block.